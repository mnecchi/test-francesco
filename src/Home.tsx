import React, { FC, useEffect, useReducer, useState } from 'react';
import {
  Alert,
  Button,
  ScrollView,
  StyleSheet,
  Text,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import fetchQuestions from './data/fetchQuestions';
import { Questions } from './data/types';
import reducer from './data/reducer';

const Home: FC = () => {
  const [questions, setQuestions] = useState<Questions>();
  const [userAnswers, dispatch] = useReducer(reducer, {});

  useEffect(() => {
    const getQuestions = async () => {
      const response = await fetchQuestions();
      setQuestions(response);
    };
    getQuestions();
  }, [setQuestions]);

  const now = new Date(Date.now()).toTimeString();

  return (
    <ScrollView>
      <Text>Render {now}</Text>
      {questions &&
        questions.map((question) => {
          const { id, text, answers } = question;
          return (
            <View testID={id} key={id} style={styles.question}>
              <Text>Render {now}</Text>
              <Text style={styles.questionTitle}>{text}</Text>
              {answers.map((answer) => {
                const { id: answerId, text: answerText } = answer;
                return (
                  <TouchableWithoutFeedback
                    testID={answerId}
                    key={answerId}
                    onPress={() => {
                      dispatch({
                        type: 'setUserAnswer',
                        payload: { id, answerId },
                      });
                    }}>
                    <Text
                      style={userAnswers[id] === answerId && styles.selected}>
                      {answerText} {Date.now()}
                    </Text>
                  </TouchableWithoutFeedback>
                );
              })}
            </View>
          );
        })}
      <View style={styles.buttonContainer}>
        <Text>Render {now}</Text>
        <Button
          testID="submit"
          disabled={
            !questions || Object.keys(userAnswers).length !== questions.length
          }
          onPress={() => {
            Alert.alert('Submit!');
          }}
          title="Submit"
        />
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  question: {
    borderColor: 'blue',
    borderWidth: 1,
    padding: 12,
    marginTop: 12,
  },
  questionTitle: {
    marginTop: 12,
    fontWeight: 'bold',
  },
  selected: {
    fontWeight: 'bold',
    color: 'green',
  },
  buttonContainer: {
    marginTop: 20,
    justifyContent: 'center',
    borderColor: 'lightblue',
    borderWidth: 1,
  },
});

export default Home;
