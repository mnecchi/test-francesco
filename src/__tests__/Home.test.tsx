import React from 'react';
import {
  act,
  create,
  ReactTestInstance,
  ReactTestRenderer,
} from 'react-test-renderer';
import { Alert } from 'react-native';
import Home from '../Home';
import mockQuestions from '../data/questions.json';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

jest
  .spyOn(global.Date, 'now')
  .mockImplementation(() => new Date('01/12/2020 12:00:00').getTime());

jest.mock('../data/fetchQuestions', () => ({
  __esModule: true,
  default: () => Promise.resolve(mockQuestions),
}));

describe('Home component', () => {
  it('should fetch and display the questions', async () => {
    let wrapper: ReactTestRenderer;

    await act(async () => {
      wrapper = create(<Home />);
    });

    expect(wrapper!.root.findByProps({ testID: 'q01' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q01a01' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q01a02' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q01a03' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q02' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q02a01' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q02a02' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q03' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q03a01' })).not.toBeUndefined();
    expect(wrapper!.root.findByProps({ testID: 'q03a02' })).not.toBeUndefined();

    expect(wrapper!).toMatchSnapshot();
  });

  it('should select and deselect an answer on click', async () => {
    let wrapper: ReactTestRenderer;

    await act(async () => {
      wrapper = create(<Home />);
    });

    const answer = wrapper!.root.findByProps({ testID: 'q01a02' });
    await act(async () => {
      answer.props.onPress();
    });
    expect(
      (answer.children[0] as ReactTestInstance).props.style.fontWeight,
    ).toBe('bold');

    await act(async () => {
      answer.props.onPress();
    });
    expect(
      (answer.children[0] as ReactTestInstance).props.style.fontWeight,
    ).toBeUndefined();
  });

  it('should enable the submit button only if all questions are answered', async () => {
    let wrapper: ReactTestRenderer;

    await act(async () => {
      wrapper = create(<Home />);
    });

    const submit = wrapper!.root.findByProps({ testID: 'submit' });
    expect(submit.props.disabled).toBe(true);

    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q01a02' }).props.onPress();
    });
    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q02a01' }).props.onPress();
    });
    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q03a02' }).props.onPress();
    });
    expect(submit.props.disabled).toBe(false);
  });

  it('should display an alert when submit is pressed', async () => {
    jest.spyOn(Alert, 'alert');
    let wrapper: ReactTestRenderer;

    await act(async () => {
      wrapper = create(<Home />);
    });
    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q01a02' }).props.onPress();
    });
    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q02a01' }).props.onPress();
    });
    await act(async () => {
      wrapper!.root.findByProps({ testID: 'q03a02' }).props.onPress();
    });

    wrapper!.root.findByProps({ testID: 'submit' }).props.onPress();
    expect(Alert.alert).toHaveBeenCalledTimes(1);
  });
});
