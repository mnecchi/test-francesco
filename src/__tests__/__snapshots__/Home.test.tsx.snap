// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should fetch and display the questions 1`] = `
<RCTScrollView>
  <View>
    <Text>
      Render 
      12:00:00 GMT+0000 (Greenwich Mean Time)
    </Text>
    <View
      style={
        Object {
          "borderColor": "blue",
          "borderWidth": 1,
          "marginTop": 12,
          "padding": 12,
        }
      }
      testID="q01"
    >
      <Text>
        Render 
        12:00:00 GMT+0000 (Greenwich Mean Time)
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginTop": 12,
          }
        }
      >
        Question 1
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q01a01"
      >
        answer 1
         
        1578830400000
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q01a02"
      >
        answer 2
         
        1578830400000
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q01a03"
      >
        answer 3
         
        1578830400000
      </Text>
    </View>
    <View
      style={
        Object {
          "borderColor": "blue",
          "borderWidth": 1,
          "marginTop": 12,
          "padding": 12,
        }
      }
      testID="q02"
    >
      <Text>
        Render 
        12:00:00 GMT+0000 (Greenwich Mean Time)
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginTop": 12,
          }
        }
      >
        Question 2
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q02a01"
      >
        answer 1
         
        1578830400000
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q02a02"
      >
        answer 2
         
        1578830400000
      </Text>
    </View>
    <View
      style={
        Object {
          "borderColor": "blue",
          "borderWidth": 1,
          "marginTop": 12,
          "padding": 12,
        }
      }
      testID="q03"
    >
      <Text>
        Render 
        12:00:00 GMT+0000 (Greenwich Mean Time)
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginTop": 12,
          }
        }
      >
        Question 3
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q03a01"
      >
        answer 1
         
        1578830400000
      </Text>
      <Text
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
        testID="q03a02"
      >
        answer 2
         
        1578830400000
      </Text>
    </View>
    <View
      style={
        Object {
          "borderColor": "lightblue",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    >
      <Text>
        Render 
        12:00:00 GMT+0000 (Greenwich Mean Time)
      </Text>
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="submit"
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            disabled={true}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
